src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfComplexTypesTests.cs

```diff
@@ -0,0 +1,176 @@
+            // at Microsoft.Spark.Worker.Command.PicklingSqlCommandExecutor.SingleCommandRunner.Run(Int32 splitId, Object input) in Microsoft.Spark.Worker\Command\SqlCommandExecutor.cs:line 239
+            // at Microsoft.Spark.Worker.Command.PicklingSqlCommandExecutor.ExecuteCore(Stream inputStream, Stream outputStream, SqlCommand[] commands) in Microsoft.Spark.Worker\Command\SqlCommandExecutor.cs:line 139
+            Func<Column, Column> udf = Udf<IDictionary<string, string>, string>(
+                    dict => dict.Count.ToString());
+            Assert.Throws<Exception>(() => _df.Select(udf(_df["info"])).Collect().ToArray());
+            Assert.Throws<Exception>(() => _df.Select(udf(_df["info"])).Show());
+        }
+
+        /// <summary>
+        /// UDF that returns Map type.
+        /// </summary>
+        [Fact]
+        public void TestUdfWithReturnTypeAsMap()
+        {
+            // UDF with return type as map throws a following exception:
+            // Unhandled Exception: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. 
+            // ---> System.NotImplementedException: The method or operation is not implemented.
+            // System.NotImplementedException: The method or operation is not implemented.
+            // at Microsoft.Spark.Sql.Row.Convert() in Microsoft.Spark\Sql\Row.cs:line 169
+            // at Microsoft.Spark.Sql.Row..ctor(Object[] values, StructType schema) in Microsoft.Spark\Sql\Row.cs:line 34
+            // at Microsoft.Spark.Sql.RowConstructor.GetRow() in MicrTosoft.Spark\Sql\RowConstructor.cs:line 113
+            // at Microsoft.Spark.Sql.RowCollector.Collect(ISocketWrapper socket) + MoveNext() in Microsoft.Spark\Sql\RowCollector.cs:line 36
+            // at Microsoft.Spark.Sql.DataFrame.GetRows(String funcName) + MoveNext() in Microsoft.Spark\Sql\DataFrame.cs:line 891
+            Func<Column, Column> udf = Udf<string, IDictionary<string, string>>(
+                str => new Dictionary<string, string> { { str, str } });
+            Assert.Throws<NotImplementedException>(() => _df.Select(udf(_df["name"])).Collect().ToArray());
+
+            //Show() works here. See the example below.
+            _df.Select(udf(_df["name"])).Show();
+        }
+
+        /// <summary>
+        /// UDF that takes in Row type.
+        /// </summary>
+        [Fact]
+        public void TestUdfWithRowType()
+        {
+            Func<Column, Column> udf = Udf<Row, string>(
+                (row) =>
+                {
+                    string city = row.GetAs<string>("city");
+                    string state = row.GetAs<string>("state");
+                    return $"{city},{state}";
+                });
+
+            Row[] rows = _df.Select(udf(_df["info"])).Collect().ToArray();

```

Comment:
You are using _df["info"] here as a Row type and also in the MapType tests.   Is it both a Row type and a Map type ?

Reply:
I have added another column with MapType.